File: src/main/java/com/derongan/minecraft/mineinabyss/gui/StatsGUI.java
package com.derongan.minecraft.mineinabyss.gui;

import com.derongan.minecraft.deeperworld.world.section.Section;
import com.derongan.minecraft.guiy.gui.*;
import com.derongan.minecraft.guiy.gui.layouts.HistoryGuiHolder;
import com.derongan.minecraft.mineinabyss.AbyssContext;
import com.derongan.minecraft.mineinabyss.MineInAbyss;
import com.derongan.minecraft.mineinabyss.configuration.ConfigConstants;
import com.derongan.minecraft.mineinabyss.player.PlayerData;
import de.erethon.headlib.HeadLib;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class StatsGUI extends HistoryGuiHolder {
    private AbyssContext context;
    private List<ClickableElement> mobConfigs = new ArrayList<>();
    private List<ClickableElement> spawnList = new ArrayList<>();
    private PlayerData playerData;
    private static ChatColor mainColor = ConfigConstants.mainColor;
    private static ChatColor secondaryColor = ConfigConstants.secondaryColor;

    public StatsGUI(Player player, MineInAbyss plugin) {
        super(6, "Mine In Abyss - Stats", plugin);
        context = MineInAbyss.getContext();
        playerData = context.getPlayerData(player);
        this.player = player;

        setElement(buildMain());
    }

    public static ItemStack getHead(Player player) {
        int lifePlayer = (int) player.getHealth();
        ItemStack item = new ItemStack(Material.PLAYER_HEAD);
        SkullMeta skull = (SkullMeta) item.getItemMeta();
        skull.setDisplayName(ChatColor.RESET + player.getName());
        skull.setOwningPlayer(player);
        item.setItemMeta(skull);
        return item;
    }

    private Layout buildMain() {
        Layout layout = new Layout();

        //Player head
        Element name = Cell.forItemStack(getHead(player), mainColor + player.getName());
        layout.addElement(0, 0, name);

        //Player's whistle level
        ItemStack whistleItem = playerData.getWhistle().getItem();
        Element whistle = Cell.forItemStack(whistleItem, mainColor + "Whisle: " + whistleItem.getItemMeta().getDisplayName());
        layout.addElement(1, 0, whistle);

        //The section the player is currently in
        Section section = context.getRealWorldManager().getSectionFor(player.getLocation());
        String layerName = "Not in a layer", sectionName = "Not in a section";
        if (section != null) {
            layerName = context.getWorldManager().getLayerForSection(section).getName();
            sectionName = section.getKey().toString().toUpperCase();
        }

        ItemStack layerHead = HeadLib.QUARTZ_L.toItemStack(mainColor + "Layer: " + secondaryColor + layerName);
        ItemMeta layerHeadMeta = layerHead.getItemMeta();
        layerHeadMeta.setLore(Arrays.asList(mainColor + "Section: " + secondaryColor + sectionName));
        layerHead.setItemMeta(layerHeadMeta);

        Element layer = Cell.forItemStack(layerHead);
        layout.addElement(2, 0, layer);

        //The player's level
        Element level = Cell.forItemStack(new ItemStack(Material.EXPERIENCE_BOTTLE), mainColor + "Level: " + ChatColor.GREEN + context.getPlayerData(player).getLevel());
        layout.addElement(7, 0, level);

        //The player's balance
        Element balance = Cell.forItemStack(new ItemStack(Material.GOLD_BLOCK), mainColor + "Balance: $" + ChatColor.GOLD + MineInAbyss.getEcon().getBalance(player));
        layout.addElement(8, 0, balance);

        addBackButton(layout);
        return layout;
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/gui/StatsGUI.kt
package com.derongan.minecraft.mineinabyss.gui

import com.derongan.minecraft.deeperworld.world.section.section
import com.derongan.minecraft.guiy.gui.ClickableElement
import com.derongan.minecraft.guiy.gui.layouts.HistoryGuiHolder
import com.derongan.minecraft.guiy.helpers.toCell
import com.derongan.minecraft.guiy.kotlin_dsl.guiyLayout
import com.derongan.minecraft.mineinabyss.MineInAbyss
import com.derongan.minecraft.mineinabyss.configuration.ConfigConstants
import com.derongan.minecraft.mineinabyss.ecs.components.PlayerData
import com.derongan.minecraft.mineinabyss.ecs.components.playerData
import com.derongan.minecraft.mineinabyss.mineInAbyss
import com.derongan.minecraft.mineinabyss.world.layer
import com.mineinabyss.geary.minecraft.access.geary
import com.mineinabyss.idofront.items.editItemMeta
import de.erethon.headlib.HeadLib
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import net.md_5.bungee.api.ChatColor as CC

class StatsGUI(private val player: Player) : HistoryGuiHolder(6, "Mine In Abyss - Stats", mineInAbyss) {
    private val col1 = ConfigConstants.mainColor
    private val col2 = ConfigConstants.secondaryColor
    private val mobConfigs: List<ClickableElement> = ArrayList()
    private val spawnList: List<ClickableElement> = ArrayList()
    private val playerData = player.playerData

    init {
        setElement(buildMain())
    }

    private fun buildMain() = guiyLayout {
        //Player head
        setElement(0, 0, getHead(player).toCell(col1.toString() + player.name))

        //The section the player is currently in
        val section = player.location.section
        val layerName = section?.let { section.layer?.name } ?: "Not in a layer"
        val sectionName = section?.key?.toString()?.uppercase() ?: "Not in a section"

        setElement(2, 0, HeadLib.QUARTZ_L.toItemStack().editItemMeta {
            setDisplayName("${col1}Layer: $col2$layerName")
            lore = listOf("${col1}Section: $col2$sectionName")
        }.toCell())

        //The player's level
        setElement(7, 0, Material.EXPERIENCE_BOTTLE.toCell("${col1}Level: ${CC.GREEN}${playerData.level}"))

        //The player's balance
        setElement(
            8, 0,
            Material.GOLD_BLOCK.toCell("${col1}Balance: ${CC.GOLD}$${MineInAbyss.econ!!.getBalance(player)}")
        )

        addBackButton(this)
    }

    private fun getHead(player: Player) = ItemStack(Material.PLAYER_HEAD).editItemMeta {
        if (this !is SkullMeta) return@editItemMeta
        setDisplayName("${CC.RESET}${player.name}")
        owningPlayer = player
    }
}
